File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {

        int carry = 0;
        ListNode res = new ListNode();
        ListNode resHead = res;

        while(l1!=null || l2!=null || carry!=0) {

            int sum = ((l1!=null)? l1.val : 0) + ((l2!=null)? l2.val : 0) + carry;
            carry = sum/10;
            int digit = sum%10;

            ListNode newNode = new ListNode(digit);
            res.next = newNode;
            res = res.next;

            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        return resHead.next;
        
    }
}

File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let carry = 0;
    let res = new ListNode(0);
    let resHead = res;


    while(l1||l2||carry) {
        let sum = (l1? l1.val : 0) + (l2 ? l2.val : 0)+ carry;
        carry = Math.floor(sum/10);
        let digit = sum%10;

        let newNode = new ListNode(digit);
        res.next = newNode;
        res = res.next

//  ---see next line---
        // what if l1 or l2 doesn't exist (means ended the size)
        l1 = l1?.next;
        l2 = l2?.next;

        // l1 = l1 && l1.next;
        // l2 = l2 && l2.next;
    }
    return resHead.next;

};